### PR TITLE
Use sendmsg rather than sendto to send UDP data.

### PR DESCRIPTION
**Pull request by PeterAdams-A**
Motivation:

sendmsg is more powerful - this is paving the way for
Explicit Congestion Notifications to be sent.

Modifications:

Replace the sendto system call with sendmsg up to DatagramChannel.

Result:

Sendmsg will be used to send UDP data.

### CODE REVIEW

File: Sources/NIO/BSDSocketAPIWindows.swift [hunk 1]
```diff
@@ -164,6 +164,13 @@ extension NIOBSDSocket {
     static func recvmsg(descriptor: CInt, msgHdr: UnsafeMutablePointer<msghdr>, flags: CInt) throws -> IOResult<ssize_t> {
         fatalError("recvmsg not yet implemented on Windows")
     }
+    
+    @inline(never)
+    static func sendmsg(descriptor: CInt,
+                        msgHdr: UnsafePointer<msghdr>,
+                        flags: CInt) throws -> IOResult<ssize_t> {
+        fatalError("recvmsg not yet implemented on Windows")
+    }
```
Comment (Lukasa) [inline Sources/NIO/BSDSocketAPIWindows.swift:173]: Please also delete `sendto` from this file.

File: Sources/NIO/SocketChannel.swift [hunk 2]
```diff
@@ -614,23 +614,29 @@ final class DatagramChannel: BaseSocketChannel<Socket> {
     override func cancelWritesOnClose(error: Error) {
         self.pendingWrites.failAll(error: error, close: true)
     }
-
+
```
Comment (Lukasa) [inline Sources/NIO/SocketChannel.swift:617]: Nit: remove this whitespace.

File: Sources/NIO/SocketChannel.swift [hunk 3]
```diff
@@ -614,23 +614,29 @@ final class DatagramChannel: BaseSocketChannel<Socket> {
...
-                                          destinationPtr: destinationPtr,
-                                          destinationSize: destinationSize)
-        }, vectorWriteOperation: { msgs in
-            try self.socket.sendmmsg(msgs: msgs)
-        })
+        )
         return result
     }
 
+
```
Comment (Lukasa) [inline Sources/NIO/SocketChannel.swift:639]: Nit: remove this whitespace.

File: Sources/NIO/Socket.swift [hunk 4]
```diff
@@ -152,21 +152,39 @@ typealias IOVector = iovec
             try Posix.writev(descriptor: $0, iovecs: iovecs)
         }
     }
-
+
```
Comment (Lukasa) [inline Sources/NIO/Socket.swift:155]: Nit: remove this whitespace.

File: Sources/NIO/Socket.swift [hunk 5]
```diff
@@ -152,21 +152,39 @@ typealias IOVector = iovec
             try Posix.writev(descriptor: $0, iovecs: iovecs)
         }
     }
-
+    
     /// Send data to a destination.
     ///
     /// - parameters:
     ///     - pointer: Pointer (and size) to the data to send.
+    ///     - controlBytes: Extra `cmsghdr` information.
```
Comment (Lukasa) [inline Sources/NIO/Socket.swift:160]: Nit: this block should match the declaration order in the code.

File: Sources/NIO/Socket.swift [hunk 6]
```diff
@@ -152,21 +152,39 @@ typealias IOVector = iovec
...
     /// - throws: An `IOError` if the operation failed.
-    func sendto(pointer: UnsafeRawBufferPointer, destinationPtr: UnsafePointer<sockaddr>, destinationSize: socklen_t) throws -> IOResult<Int> {
+    func sendmsg(pointer: UnsafeRawBufferPointer,
+                 destinationPtr: UnsafePointer<sockaddr>,
+                 destinationSize: socklen_t,
+                 controlBytes: UnsafeMutableRawBufferPointer) throws -> IOResult<Int> {
+        var vec = iovec(iov_base: UnsafeMutableRawPointer(mutating: pointer.baseAddress!), iov_len: pointer.count)
+        
+        // Dubious const cast - it should be OK as there is no reason why this should get mutated
+        // just bad const declaration below us.
```
Comment (Lukasa) [inline Sources/NIO/Socket.swift:173]: This is really a pair of dubious casts: want to lift this comment up to cover both lines?

File: Sources/NIO/Socket.swift [hunk 7]
```diff
@@ -152,21 +152,39 @@ typealias IOVector = iovec
...
+        let notConstCorrectDestinationPtr = UnsafeMutableRawPointer(mutating: destinationPtr)
+
         return try withUnsafeHandle {
-            try NIOBSDSocket.sendto(socket: $0,
-                                    buffer: UnsafeMutableRawPointer(mutating: pointer.baseAddress!),
-                                    length: pointer.count,
-                                    dest_addr: destinationPtr,
-                                    dest_len: destinationSize)
+            let handle = $0
```
Comment (Lukasa) [inline Sources/NIO/Socket.swift:177]: Amend the above line to `return try withUnsafeHandle { handle in` instead of using this.

**Comment by PeterAdams-A** [comment]
Suggested review changes made.

**Pull request merged**